### PR TITLE
Fix the History api

### PR DESCRIPTION
**Pull request by falsandtru**
Fixes https://github.com/Microsoft/TypeScript/issues/10059

### CODE REVIEW

**Comment by falsandtru** [comment]
@zhengbli How can I remove `declare var History`?

**Comment by zhengbli** [comment]
Why do you need to remove the `declare var History`? We need a constructor for the `typeof` operator to work, also according to https://github.com/Microsoft/TypeScript/issues/10059#issuecomment-236814714, `console.log(History)` does print something in Chrome.

**Comment by falsandtru** [comment]
Yes, I fixed it. Function signatures are needed for the `instanceof` function as follows.

```ts
history instanceof History
```

**Comment by falsandtru** [comment]
Ah, no, removed.

**Comment by falsandtru** [comment]
`history` is defined in the following spec, but `History` is not.

https://html.spec.whatwg.org/multipage/browsers.html#the-window-object

**Comment by zhengbli** [comment]
The constructor is indeed defined in major browsers, it is just forbidden to call. For comparability I would suggest just leave it there. @mhegazy thoughts?
capture
capture1

**Comment by falsandtru** [comment]
I think `History` variable should be defined by users because it doesn't standardized by spec.

File: baselines/dom.generated.d.ts [hunk 1]
```diff
@@ -7536,16 +7536,12 @@ declare var HashChangeEvent: {
...
     readonly state: any;
-    back(distance?: any): void;
-    forward(distance?: any): void;
-    go(delta?: any): void;
-    pushState(statedata: any, title?: string, url?: string): void;
-    replaceState(statedata: any, title?: string, url?: string): void;
-}
-
-declare var History: {
-    prototype: History;
```
Comment (mhegazy) [inline baselines/dom.generated.d.ts]: there is a var `History` i find in most browsers. why was this removed?

Comment (falsandtru) [inline baselines/dom.generated.d.ts]: See https://github.com/Microsoft/TSJS-lib-generator/pull/142#issuecomment-242893870

Comment (mhegazy) [inline baselines/dom.generated.d.ts]: The `History` is the prototype for the type. the constructor can not be called directly, but that is the same as the rest of the HTML*Element classes as well. so we should keep both a type and a value for `History`. we already have a value `history: History` on `Window` and the global object.

Comment (falsandtru) [inline baselines/dom.generated.d.ts]: Okay, fixed.

File: baselines/dom.generated.d.ts [hunk 2]
```diff
@@ -7536,11 +7536,12 @@ declare var HashChangeEvent: {
...
-    back(distance?: any): void;
-    forward(distance?: any): void;
-    go(delta?: any): void;
-    pushState(statedata: any, title?: string, url?: string): void;
-    replaceState(statedata: any, title?: string, url?: string): void;
+    scrollRestoration: ScrollRestoration;
+    back(): void;
+    forward(): void;
+    go(delta?: number): void;
+    pushState(data: any, title: string, url?: string | null): void;
```
Comment (zhengbli) [inline baselines/dom.generated.d.ts]: optional parameter doesn't need the `| null` annotation, it gets that automatically.
Also, if missing the value of the optional parameter should be `undefined` I think

File: baselines/dom.generated.d.ts [hunk 3]
```diff
@@ -7536,11 +7536,12 @@ declare var HashChangeEvent: {
 interface History {
     readonly length: number;
     readonly state: any;
-    back(distance?: any): void;
-    forward(distance?: any): void;
-    go(delta?: any): void;
-    pushState(statedata: any, title?: string, url?: string): void;
-    replaceState(statedata: any, title?: string, url?: string): void;
+    scrollRestoration: ScrollRestoration;
```
Comment (zhengbli) [inline baselines/dom.generated.d.ts]: is it a read only property? If so add the modifier

Comment (falsandtru) [inline baselines/dom.generated.d.ts]: See https://html.spec.whatwg.org/multipage/browsers.html#the-history-interface

File: baselines/dom.generated.d.ts (hunk 2 above)
Comment (falsandtru) [inline baselines/dom.generated.d.ts]: Its signature in spec is `void pushState(any data, DOMString title, optional USVString? url = null);`. So it needs to accept a `null` value and should be optional.

https://html.spec.whatwg.org/multipage/browsers.html#the-history-interface

Comment (zhengbli) [inline baselines/dom.generated.d.ts]: Ah, makes sense. Thanks

**Comment by zhengbli** [comment]
👍 Thanks!
